resources/assets/lib/user-list.tsx

```diff
@@ -0,0 +1,170 @@
+}
+
+interface State {
+  filter: Filter;
+  sortMode: SortMode;
+}
+
+export class UserList extends React.PureComponent<Props> {
+  readonly state: State = {
+    filter: this.filterFromUrl,
+    sortMode: this.sortFromUrl,
+  };
+
+  onSortSelected = (event: React.SyntheticEvent) => {
+    const value = (event.currentTarget as HTMLElement).dataset.value;
+    const url = osu.updateQueryString(null, { user_sort: value });
+
+    Turbolinks.controller.pushHistoryWithLocationAndRestorationIdentifier(url, Turbolinks.uuid());
+    this.setState({ sortMode: value });
+  }
+
+  optionSelected = (event: React.SyntheticEvent) => {
+    event.preventDefault();
+    const key = (event.currentTarget as HTMLElement).dataset.key;
+    const url = osu.updateQueryString(null, { filter: key });
+
+    Turbolinks.controller.pushHistoryWithLocationAndRestorationIdentifier(url, Turbolinks.uuid());
+    this.setState({ filter: key });
+  }
+
+  render(): React.ReactNode {
+    return (
+      <>
+        {this.renderSelections()}
+        <div className='user-list'>
+          <div className='user-list__toolbar'>
+            {this.renderSorter()}
+          </div>
+
+          <UserCards users={this.sortedUsers} />
+        </div>
+      </>
+    );
+  }
+
+  renderSelections() {
+    const groups = [
+      { key: 'all', count: this.props.users.length },
+      { key: 'online', count: this.props.users.filter((user) => user.is_online).length },
+    ];
+
+    return (
+      <div className='update-streams-v2 update-streams-v2--with-active'>
+        <div className='update-streams-v2__container'>
+          {
+            groups.map((group) => {
+              return this.renderOption(group.key, group.count, group.key === this.state.filter);
+            })
+          }
+        </div>
+      </div>
+    );
+  }
+
+  renderSorter() {
+    return (
+      <Sort
+        modifiers={['user-list']}
+        onSortSelected={this.onSortSelected}
+        sortMode={this.state.sortMode}
+        values={sortModes}
+      />
+    );
+  }
+
+  renderOption(key: string, text: string | number, active = false) {
+    // FIXME: change all the names
+    const modifiers = active ? ['active'] : [];
+    let className = osu.classWithModifiers('update-streams-v2__item', modifiers);
+    className += ` t-changelog-stream--${key}`;
+
+    return (
+      <a
+        className={className}
+        data-key={key}
+        href={osu.updateQueryString(null, { filter: key })}
+        key={key}
+        onClick={this.optionSelected}
+      >
+        <div className='update-streams-v2__bar u-changelog-stream--bg' />
+        <p className='update-streams-v2__row update-streams-v2__row--name'>{osu.trans(`users.status.${key}`)}</p>
+        <p className='update-streams-v2__row update-streams-v2__row--version'>{text}</p>
+      </a>
+    );
+  }
+
+  private get sortedUsers() {
+    const users = this.filteredUsers.slice();
+
+    switch (this.state.sortMode) {
+      case 'username':
```

Comment:
More like `case 'last_visit'`.

Reply:
probably switch the two return statements instead, does the same thing but would make clear which one is the default